File: evals/evals.ts
import { LLMClassifierFromTemplate,Faithfulness } from "autoevals";
import { Eval, projects, initDataset } from "braintrust";
import { OpenAI } from "openai";
import { Pinecone } from "@pinecone-database/pinecone";
import { VoyageAIClient } from "voyageai";
import { wrapTraced, currentSpan } from "braintrust";

// Setup Pinecone, VoyageAI, and OpenAI clients
const pinecone = new Pinecone({ apiKey: process.env.PINECONE_API_KEY! });
const index = pinecone.Index("braintrust-rag-bot");

const voyage = new VoyageAIClient({
  apiKey: process.env.VOYAGEAI_API_KEY,
});
const MODEL = "voyage-3";

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

interface Input {
  content: string;
  parts: object[];
  role: string;
}

// Create embedding and retrieval functions
const embedQuery = wrapTraced(async function embedQuery(query: Input[]) {
  console.log("EMBEDDING QUERY", query[0].content);
  const docs = await voyage.embed({ input: query[0].content, model: MODEL });
  return docs.data?.[0]?.embedding ?? [];
});

export const getDocs = wrapTraced(
  async function getDocs(query: Input[]) {
    const embedding = await embedQuery(query);

    const results = await index.query({
      vector: embedding,
      topK: 3,
      includeMetadata: true,
    });

    const context =
      results.matches?.map((match) => match.metadata?.content).join("\n\n") ??
      "";
    currentSpan().log({
      metadata: {
        context: context,
      },
    });
    return context;
  },
  { type: "tool" }
);

async function getOutput(query: Input[]): Promise<Output> {
  const context = await getDocs(query);
  const response = await openai.chat.completions.create({
    model: "gpt-4o-mini",
    messages: [
      {
        role: "system",
        content: `
      You are a helpful assistant that can answer questions about Braintrust using the Braintrust documentation.
      `,
      },
      { role: "user", content: query[0].content },
      {
        role: "assistant",
        content: `
      CONTEXT: ${context}
      `,
      },
    ],
  });

  const output: Output = {output: response.choices[0].message.content ?? "", context: context, input: query[0].content}
  return output
}

// async function getFaithfulness(args: { input: string }) {
//   const output_obj = JSON.parse(args.input);
//   const output = output_obj.output;
//   const context = output_obj.context;
//   const classifier = LLMClassifierFromTemplate({
//     name: "NoHallucination",
//     model: "gpt-4o-mini",
//     promptTemplate: `You are an LLM editor that ensures that no false information appears in a response. Given a response
//     check the response against the context. If there is information in the response that was not in the context, this is a 
//     hallucination. If there is no hallucination, return "yes". If there is a hallucination, return "no".
    
//     Response: ${output}
    
//     Context: ${context}
//     `,
//     choiceScores: {
//       "yes": 1,
//       "no": 0,
//     },
//   });
//   return classifier
// }

interface Output {
  output: string;
  context: string;
  input: string;
}
 
const getFaithfulness = (args: {
  output: Output
}) => {
  return Faithfulness({
    output: args.output.output,
    context: args.output.context,
    input: args.output.input
  });
};

Eval("PhilScratchArea", {
  task: getOutput,
  data: initDataset({ project: "PhilScratchArea", dataset: "RAGDataset" }), // ignored
  scores: [getFaithfulness],
});

File: python_assets/evals.py
import braintrust
from pinecone.grpc import PineconeGRPC as Pinecone
from pydantic import BaseModel
import voyageai
import os
from typing import List
from openai import OpenAI
from dotenv import load_dotenv
import json
from autoevals import Faithfulness
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

load_dotenv( '../.env.local'  )

logger.info("Initializing project and clients")
project = braintrust.projects.create("PhilScratchArea")
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

class Args(BaseModel):
    query: str

class Document(BaseModel):
    content: str


class DocumentOutput(BaseModel):
    text: List[Document]

def handler(query: str):
    logger.info(f"Processing query: {query}")
    query = query[0]['content']
    logger.debug(f"Parsed query content: {query}")
    
    logger.info("Initializing VoyageAI and Pinecone clients")
    vo = voyageai.Client(api_key=os.getenv("VOYAGEAI_API_KEY"))
    pc = Pinecone(api_key=os.getenv("PINECONE_API_KEY"))
    index = pc.Index("braintrust-rag-bot")
    MODEL = 'voyage-3'

    logger.info("Generating embeddings and querying Pinecone")
    xq = vo.embed(query, model=MODEL, input_type='query').embeddings[0]
    response = index.query(xq, top_k=3, include_metadata=True)
    matches = [{'content': match['metadata']['content']} for match in response['matches']]
    logger.info(f"Found {len(matches)} matching documents")

    return matches

def prompt(query, context: list[dict]): 
    logger.info("Generating response with context")
    logger.info(f"Query: {query}")
    query = query[0]['content']
    all_content = '\n\n'.join([match['content'] for match in context])
    logger.debug(f"Combined context length: {len(all_content)} characters")
    
    logger.info("Calling OpenAI API")
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[
            {"role": "system", "content": "You are a helpful assistant that can answer questions about the Braintrust documentation. You will receive a question and context about that question. Use the context to answer the question. If the context does not contain the answer, say so."},
            {"role": "user", "content": query},
            {"role": "user", "content": f"Context: {all_content}"}
        ]
    )
    logger.info("Received response from OpenAI")

    return {"answer": response.choices[0].message.content, "context": all_content}

example_prompt = """
[
  {
    "content": "What are the different types of scorers I can use?",
    "parts": [
      {
        "text": "What are the different types of scorers I can use?",
        "type": "text"
      }
    ],
    "role": "user"
  }
]
"""

def get_output(query):
    logger.info("Starting get_output function")
    logger.info(f"Query: {query}")
    context = handler(query)
    result = prompt(query, context)
    logger.info("Completed get_output function")
    return result


def get_faithfulness(output, **kwargs):
    logger.info("Calculating faithfulness score")
    score = Faithfulness().eval(output=output['answer'], context=output['context'], **kwargs)
    logger.info(f"Faithfulness score calculated: {score}")
    return score

logger.info("Starting evaluation")
braintrust.Eval("PhilScratchArea",
                task=get_output,
                scores=[get_faithfulness],
                data= braintrust.init_dataset(project='PhilScratchArea', name='RAGDataset')
                )
logger.info("Evaluation completed")
